src/state_machine.zig

```diff
@@ -552,19 +537,22 @@ pub fn StateMachineType(
                 },
                 inline else => |comptime_operation| {
                     const event_size = @sizeOf(Event(comptime_operation));
-                    const batch_max = comptime constants.batch_max.operation_batch_max(
-                        comptime_operation,
-                    );
-
+                    const result_size = @sizeOf(Result(comptime_operation));
                     comptime assert(event_size > 0);
-                    comptime assert(batch_max > 0);
+                    comptime assert(result_size > 0);
+
+                    const batch_limit = @min(
+                        @divFloor(self.batch_size_limit, event_size),
+                        @divFloor(self.batch_size_limit, result_size),
+                    );
+                    assert(batch_limit > 0);
 
                     // Clients do not validate batch size == 0,
                     // and even the simulator can generate requests with no events.
                     maybe(input.len == 0);
 
                     if (input.len % event_size != 0) return false;
-                    if (input.len > batch_max * event_size) return false;
+                    if (input.len > batch_limit * event_size) return false;
```

Comment:
Can we simplify this to just `input.len > self.batch_size_limit`?

Reply:
Ah, no, because we also look at the result set. 

:thinking: so this is actually tricky, this means that the client generates a request, they need to concern themselves not only with the maximum size of input, but also with the maximum size of output. 

Lets maybe simplify this and:

* add comptime check that sizeOf(Result) <= sizeOf(Event)
* add a runtime check in commit that the result is not larger than we expect